File: backend/routers/vehicles.py
from typing import List
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from database import get_db
from models import Vehicle, User
from schemas.schemas import VehicleCreate, VehicleUpdate, VehicleOut
from utils.auth import get_current_user

router = APIRouter(prefix="/api/vehicles", tags=["vehicles"])


@router.get("", response_model=List[VehicleOut])
def list_vehicles(db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    return db.query(Vehicle).all()


@router.post("", response_model=VehicleOut, status_code=status.HTTP_201_CREATED)
def create_vehicle(payload: VehicleCreate, db: Session = Depends(get_db),
                    current_user: User = Depends(get_current_user)):
    existing = db.query(Vehicle).filter(Vehicle.vehicle_number == payload.vehicle_number).first()
    if existing:
        raise HTTPException(status_code=400, detail="Vehicle number already exists")
    vehicle = Vehicle(**payload.model_dump())
    db.add(vehicle)
    db.commit()
    db.refresh(vehicle)
    return vehicle


@router.put("/{vehicle_id}", response_model=VehicleOut)
def update_vehicle(vehicle_id: int, payload: VehicleUpdate, db: Session = Depends(get_db),
                    current_user: User = Depends(get_current_user)):
    vehicle = db.query(Vehicle).get(vehicle_id)
    if not vehicle:
        raise HTTPException(status_code=404, detail="Vehicle not found")
    data = payload.model_dump(exclude_unset=True)
    for field, value in data.items():
        setattr(vehicle, field, value)
    db.commit()
    db.refresh(vehicle)
    return vehicle

File: frontend/src/api/vehicles.js
import api from "./axios";

// GET /api/vehicles -> VehicleOut[]
export function listVehicles() {
  return api.get("/api/vehicles").then((r) => r.data);
}

// POST /api/vehicles -> VehicleOut
// payload: { vehicle_number, vehicle_type, capacity_kg }
export function createVehicle(payload) {
  return api.post("/api/vehicles", payload).then((r) => r.data);
}

// PUT /api/vehicles/{id} -> VehicleOut
// payload: { status?, fuel_level?, current_lat?, current_lng? }
export function updateVehicle(id, payload) {
  return api.put(`/api/vehicles/${id}`, payload).then((r) => r.data);
}
